plugin/pulpcore/plugin/repository.py

```diff
@@ -0,0 +1,171 @@
+
+        Returns:
+            pulpcore.plugin.repository.RepositoryVersion: The Created RepositoryVersion
+        """
+        assert get_current_task_id() is not None, _('RepositoryVersion creation must be ran inside'
+                                                    ' a task')
+        with transaction.atomic():
+            version = models.RepositoryVersion(
+                repository=repository,
+                number=repository.last_version + 1)
+            repository.last_version = version.number
+            repository.save()
+            version.save()
+            resource = models.CreatedResource(content_object=version)
+            resource.save()
+            return cls(version, resource)
+
+    def __init__(self, model, resource=None):
+        """
+        Args:
+            model (pulpcore.models.repository.RepositoryVersion)
+            resource (pulpcore.models.task.CreatedResource)
+        """
+        self._model = model
+        self._created_resource = resource
+
+    @property
+    def number(self):
+        return self._model.number
+
+    @property
+    def content(self):
+        """
+        QuerySet of content on this repository version
+
+        Returns:
+            QuerySet: The Content objects that are related to this version.
+        """
+        return self._model.content
+
+    def add_content(self, content):
+        """
+        Add a content unit to the repository.
+
+        Args:
+           content (pulpcore.plugin.models.Content): a content model to add
+        """
+        assert not self._model.complete, _("Cannot update version {version} for repository {repo}."
+                                           " Complete Repository is immutable")\
+            .format(version=self._model.number, repo=self._model.repository.name)
+
+        with suppress(IntegrityError):
+            association = models.RepositoryContent(
+                repository=self._model.repository,
+                content=content,
+                version_added=self._model)
+            association.save()
+
+    def remove_content(self, content):
+        """
+        Remove content from the repository.
+
+        Args:
+            content (pulpcore.plugin.models.Content): A content model to remove
+        """
+        assert not self._model.complete, _("Cannot update version {version} for repository {repo}."
+                                           " Complete Repository is immutable")\
+            .format(version=self._model.number, repo=self._model.repository.name)
+
+        q_set = models.RepositoryContent.objects.filter(
+            repository=self._model.repository,
+            content=content,
+            version_removed=None)
+        q_set.update(version_removed=self._model)
+
+    def complete(self):
+        """
+        Save the repository version when all operations are done on it
+        """
+        with transaction.atomic():
```

Comment:
Why is this in a transaction?

Reply:
Whoops, this does not need to be, will remove